File: src/main/java/de/pflugradts/pwman3/domain/model/transfer/Chars.java
package de.pflugradts.pwman3.domain.model.transfer;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.RequiredArgsConstructor;
import lombok.ToString;

import java.security.SecureRandom;

import static de.pflugradts.pwman3.domain.model.transfer.CharValue.MAX_ASCII_VALUE;
import static de.pflugradts.pwman3.domain.model.transfer.CharValue.MIN_ASCII_VALUE;

/**
 * <p>A Chars is used to represent data send to or received from external interfaces that uses char[].</p>
 * <p>A Chars can be constructed from and converted to a {@link Bytes}
 * which is basic structure used to represent data.</p>
 */
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
@EqualsAndHashCode
@ToString
public class Chars {

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    private final char[] charArray;

    public static Chars of(final char... chars) {
        return new Chars(chars);
    }

    public static void scramble(char... chars) {
        for (int i = 0; i < chars.length; i++) {
            chars[i] = (char) (SECURE_RANDOM.nextInt(1 + MAX_ASCII_VALUE - MIN_ASCII_VALUE) + MIN_ASCII_VALUE);
        }
    }

    public Bytes toBytes() {
        final var byteArray = new byte[charArray.length];
        for (int i = 0; i < charArray.length; i++) {
            byteArray[i] = (byte) charArray[i];
        }
        scrambleSelf();
        return Bytes.of(byteArray);
    }

    public void scrambleSelf() {
        scramble(charArray);
    }

    public char[] toCharArray() {
        return charArray.clone();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/transfer/Chars.kt
package de.pflugradts.passbird.domain.model.transfer

import de.pflugradts.passbird.domain.model.transfer.Bytes.Companion.bytesOf
import java.security.SecureRandom

class Chars private constructor(private val charArray: CharArray) {

    fun scramble() = charArray.indices.forEach {
        charArray[it] = (SECURE_RANDOM.nextInt(1 + MAX_ASCII_VALUE - MIN_ASCII_VALUE) + MIN_ASCII_VALUE).toChar()
    }

    fun toBytes(): Bytes = bytesOf(
        ByteArray(charArray.size).also { byteArray ->
            charArray.indices.forEach { index -> byteArray[index] = charArray[index].code.toByte() }
        },
    ).also { this.scramble() }

    fun toCharArray() = charArray.clone()

    override fun equals(other: Any?): Boolean = when {
        (this === other) -> true
        (javaClass != other?.javaClass) -> false
        else -> charArray contentEquals (other as Chars).charArray
    }

    override fun hashCode() = charArray.contentHashCode()

    companion object {
        val SECURE_RANDOM = SecureRandom()

        @JvmStatic
        fun charsOf(chars: CharArray) = Chars(chars.clone())
    }
}
